File: cpp/145.Binary Tree Postorder Traversal.cpp
/**
 * Source: LeetCode  145. Binary Tree Postorder Traversal
 * Author: ceezyyy
 * Date: 10-04-2019
 */


 /*
 Hard

 Given a binary tree, return the postorder traversal of its nodes' values.

 Example:

 Input: [1,null,2,3]
    1
	 \
	  2
	 /
    3 

 Output: [3,2,1]

 Follow up: Recursive solution is trivial, could you do it iteratively?

 */


#include<vector>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	vector<int> res;
	vector<int> postorderTraversal(TreeNode* root) {
		postorderTraversal_Recursion(root);
		return res;
	}

	void postorderTraversal_Recursion(TreeNode* node) {
		if (node) {
			if (node->left)
				postorderTraversal_Recursion(node->left);
			if (node->right)
				postorderTraversal_Recursion(node->right);
			res.push_back(node->val);
		}
	}
};

/**
 * Runtime: 4 ms, faster than 58.08% of C++ online submissions for Binary Tree Postorder Traversal.
 * Memory Usage: 9.6 MB, less than 16.13% of C++ online submissions for Binary Tree Postorder Traversal.
 */

File: Python3/145.Binary Tree Postorder Traversal.py
"""
@Title: 145. Binary Tree Postorder Traversal
@Tag: binary tree
@Date: Mar-05 2020
@Author: ceezyyy
@Difficulty: Hard
"""


# Approach 1: (Iterative)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        res, stack = [], [(root, False)]
        while stack:
            node, visited = stack.pop()  # the last element
            if node:
                if visited:
                    res.append(node.val)
                else:  # postorder: left -> right -> root
                    stack.append((node, True))
                    stack.append((node.right, False))
                    stack.append((node.left, False))
        return res


"""
Runtime: 24 ms, faster than 87.85% of Python3 online submissions for Binary Tree Postorder Traversal.
Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Binary Tree Postorder Traversal.
"""


"""
Time Complexity: O(n)
Space Complexity: O(n)
"""


# Approach 2: (Recursive)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        self.dfs(root, res)
        return res

    def dfs(self, node, res):
        if node:  # left -> right -> root
            self.dfs(node.left, res)
            self.dfs(node.right, res)
            res.append(node.val)


"""
Runtime: 24 ms, faster than 87.75% of Python3 online submissions for Binary Tree Postorder Traversal.
Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Binary Tree Postorder Traversal.
"""


"""
Time Complexity: O(n)
Space Complexity: O(n)
"""
